### ISSUE TITLE
SIGSEGV handling inside Go routine

### ISSUE THREAD

**Issue by SfnUser** [onsi/ginkgo#1079]
**Setup:**
Go: v1.18.2
Ginkgo:v2.4.0

**Test Invocation**
./testsuite-name.test --ginkgo.vv --ginkgo.timeout 10s --ginkgo.fail-fast

**Problem**
In a Go routine, I have added `defer GinkgoRecover()` at the start and if there is a code that causes SIGSEGV, the test does not seem to stop immediately (depending on what the test does) but it waits for the timeout to expire. 
```
go func() {
    defer GinkgoRecover()
    var i *int
    *i = 10
    ... do something...
}()
```

Is there a way for the test to stop immediately in case of a panic like SIGSEGV and not to wait till the timeout period or run further cleanup routines? Because, if I don't have `defer GinkgoRecover()` in the Go routine, it fails immediately with SIGSEGV & stack trace, which seems straightforward. 

Moreover, in case of timeout ending the tests, need to look at all Go routine stack traces + any further panics in AfterEach etc. In large integration test suites, this is time consuming.

Any thoughts please?

**Comment by onsi** [comment]
Hey there @SfnUser 

So normally if `defer GinkgoRecover()` isn’t present this goroutine will panic and cause the entire test suite to end immediately.  This will prevent any cleanup code from running for this test _and_ it will prevent any subsequent tests from running.  The panic can occur either because of a `SIGSEGV` like this one - or, simply, because of a failure in the goroutine (ginkgo’s `Fail` function emits a panic).

I’m assuming your code is hanging after the goroutine panics because something in your spec is waiting for the goroutine to exit.  Things _eventually_ end when the `—ginkgo.timeout` (which is a suite-wide timeout) elapses.

So, for example, this:

```go
	It("sigsegvs", func(ctx SpecContext) {
		c := make(chan struct{})
		go func() {
			defer GinkgoRecover()
                        Expect(true).To(BeFalse()) // this fails, and in doing so panics, and so close(c) is never called
			close(c)
		}()
                <-c
	})
```

Will block forever (or until the suite timeout forcibly ends things).  Note that this has nothing to do with the nature of the panic (`SIGSEGV` or otherwise). It’ just how this particular spec was designed.  Rather than rely on a global timeout we can apply a local one:

```go
	It("sigsegvs", func() {
		c := make(chan struct{})
		go func() {
			defer GinkgoRecover()
                        Expect(true).To(BeFalse()) // this fails, and in doing so panics, and so close(c) is never called
			close(c)
		}()
                Eventually(c).Within(time.Second*10).Should(BeClosed())
	})
```

Now you’ll get a much more focused error message about the specific test in question, and the rest of your suite won’t hang.  This also better captures, more locally, the semantics of what you’re trying to test: “this thing is off happening in a goroutine, i have no idea if it will work or not, but eventually this channel should be closed.”

Alternatively, you can use the new [support for node timeouts and interruptible nodes](https://onsi.github.io/ginkgo/#spec-timeouts-and-interruptible-nodes):

```go
	It("sigsegvs", func(ctx SpecContext) {
		c := make(chan struct{})
		go func() {
			defer GinkgoRecover()
                        Expect(true).To(BeFalse()) // this fails, and in doing so panics, and so close(c) is never called
			close(c)
		}()
                Eventually(ctx, c).Should(BeClosed())
	}, NodeTimeout(time.Second*10))
```

Or even just:

```go
	It("sigsegvs", func(ctx SpecContext) {
		c := make(chan struct{})
		go func() {
			defer GinkgoRecover()
                        Expect(true).To(BeFalse()) // this fails, and in doing so panics, and so close(c) is never called
			close(c)
		}()
                <-c //the original code!
	}, NodeTimeout(time.Second*10), GracePeriod(time.Second))
```
This last example works but is not ideal.  After 10 seconds Ginkgo will mark the It as timedout and cancel its context.  However since nothing is monitoring the context the It will continue to hang.  Eventually, after the grace period of 1s, Ginkgo will move on to the cleanup nodes and **leak** the running It, which will continue running/hanging in the background while other tests run.  Often times that’s ok, but in more complex suites it can lead to confusing behavior, so it’s always better to write your tests in such a way that they will exit cleanly when the context is cancelled.

**Comment by onsi** [comment]
As for this:
> Moreover, in case of timeout ending the tests, need to look at all Go routine stack traces + any further panics in AfterEach etc. In large integration test suites, this is time consuming.

I assume this is referring to the case where there is a large test suite with many tests running in parallel.  If the suite level timeout is causing many tests to timeout (instead of just one or two stuck ones) then that is a sign that the timeout is too stringent and should be eased.  By combining a generous “just in case” timeout at the suite level (i.e. `—ginkgo.timeout`) with more tailored `SpecTimeout` and `NodeTimeout` decorators at the spec and node level, you should be able to build a test suite that is resilient to this issue and gives you helpful timeout feedback when individual timeouts occur.

BTW all this `SpecContext` and `SpecTimeout`/`NodeTimeout`/`GracePeriod` stuff is relatively new - take a look at the docs i linked to above!

**Comment by SfnUser** [comment]
Hi @onsi,

Apologies for the delayed response & thank you for the detailed suggestions. As you suggested, the combination of timeouts `NodeTimeout`,`SpecTimeout` does help to narrow down the failing spec.

However, in my scenario, I have multiple Go routines in a spec and I use `sync.WaitGroup` to synchronise them at the end of the spec using `wg.Wait`. If one of the Go routines crash, even with a `NodeTimeout` expiry, it does not print out the code statement that causes the SIGSEGV. It prints the stack trace of the Go routines in the spec (not much useful) and then it waits for the global timeout to expire. Finally, in the end it prints the panic causing code statement with stack trace.

While I do get the root cause at the end, I am wondering if there is a way to exit immediately or print the violating statement in the stack trace when the `Node or SpecTimeout` expires. Moreover, for a `NodeTimeout` value of 2 secs and a global timeout `--ginkgo.timeout 10s`, the test expires after ~32s - not sure why.

Is there anyway to use the `ctx` and break out of `wg.Wait` when there is a crash & when the `NodeTimeout` or `SpecTimeout` expires?

Many thanks.

**Comment by onsi** [comment]
> I am wondering if there is a way to exit immediately or print the violating statement.

There is not - the panic condition is a test failure and Ginkgo is trying to capture that failure to present it to you as part of the spec's state and then continue on to the next spec.  It is trying to avoid aborting the suite with the panic.  I would recommend continuing to invest in improving the code so that a cancelled context causes it to exit more cleanly so that Ginkgo can report the error then move on.

> Is there anyway to use the ctx and break out of wg.Wait when there is a crash & when the NodeTimeout or SpecTimeout expires?

There's no way to break out of the `wg.Wait` but I believe the correct way to solve this is to pass the `ctx` to each of your goroutines.    Then when the context is cancelled the Goroutines should finish up and the wg.Wait should return.  If you can point me at a repo with the code in question I can try to help show how to do this.

> Moreover, for a NodeTimeout value of 2 secs and a global timeout --ginkgo.timeout 10s, the test expires after ~32s - not sure why.

That is because of the `GracePeriod` documented [here](https://onsi.github.io/ginkgo/#mental-model-the-life-cycle-of-interruptions-and-the-graceperiod-decorator).  When the timeout elapses Ginkgo cancels the context.  It then needs to wait for the code under test to respond to the cancelled context and cleanup/finish up.  Ginkgo waits for the `GracePeriod` for that to happen.  If it simply continued without waiting then there would not be enough time for your code to clean upon response to the cancelled context.    I think once you pass the `ctx` down to your goroutines and have them exit appropriately when the context is cancelled you will avoid this issue.  When the timeout elapses the context will be cancelled and your goroutines will exit soon after and then your `wg.Wait` will soon exit and then Ginkgo will log the failure and move on to the other specs.

Again - happy to look at some code and help if that would be possible.

**Comment by SfnUser** [comment]
Thanks @onsi for the suggestions. Playing with the `GracePeriod` config helped resolve my scenario.

Here is an example below:

```
func RoutineGood() {
	log.Printf("all good")
	time.Sleep(time.Second*1)
}
func RoutineCrash() {
	var i *int
	*i = 10
}

// var _ = BeforeSuite(func() {})
// var _ = AfterSuite(func() {})
var _ = Describe("test crash", func() {
	var wg *sync.WaitGroup
	//BeforeEach(func() {})
	//AfterEach(func() {})
	Context("case 1", func() {
		It("crash", func(ctx SpecContext) {
			wg = &sync.WaitGroup{}
			wg.Add(1)
			go func() {
				defer GinkgoRecover()
				RoutineCrash()
				wg.Done()
			}()
			Eventually(func() bool {
				wg.Wait()
				log.Printf("Eventually exiting")
				return true
			}).WithContext(ctx).Should(BeTrue())
		}, NodeTimeout(time.Second*2))

		It("no crash", func(ctx SpecContext) {
			wg = &sync.WaitGroup{}
			wg.Add(1)
			go func() {
				defer GinkgoRecover()
				RoutineGood()
				wg.Done()
			}()
			Eventually(func() bool {
				wg.Wait()
				log.Printf("Eventually done")
				return true
			}).WithContext(ctx).Should(BeTrue())
		}, NodeTimeout(time.Second*6))
	})
})
```
and I am invoking the test as below:

```
./crash.test --ginkgo.vv --ginkgo.trace --ginkgo.timeout 20s --ginkgo.grace-period 2s
```

If there is a test condition failure, `GinkgoRecover` should print it then and there.
If there is a test case getting stuck because of logic (no runtime error), the `NodeTimeout` should kickin and print the stack trace of all routines and that should give a hint. Eventually, after the global `GracePeriod` the test will exit. I can control `GracePeriod` on a per-node basis so that other `good` tests can proceed with the `leak` though. Test writer's choice.
If there is a runtime violation, `NodeTimeout` will show the`Spec/Node` causing issue and after the `GracePeriod` it should print the stack trace causing panic. All tests will understandably stop when runtime violation happens.

With this, I don't have to play with the global test timeout to find the violating code line.

Is this understanding correct?

**Comment by onsi** [comment]
Hey @SfnUser 

yes this is moving in the right direction but here's a version that will exit shortly after the timeout occurs without needing to wait for the grace period to expire:

```go

func RoutineGood(ctx context.Context) {
	log.Printf("all good")
   select {
       <-time.After(time.Second):
       <-ctx.Done():
   }
}
func RoutineCrash(ctx context.Context) {
   //in reality your code would also monitor ctx.Done() and exit early
	var i *int
	*i = 10
}

// var _ = BeforeSuite(func() {})
// var _ = AfterSuite(func() {})
var _ = Describe("test crash", func() {
	var wg *sync.WaitGroup
	//BeforeEach(func() {})
	//AfterEach(func() {})
	Context("case 1", func() {
		It("crash", func(ctx SpecContext) {
			wg = &sync.WaitGroup{}
			wg.Add(1)
			go func() {
				defer GinkgoRecover()
            defer wg.Done() //always tell the wait group we are done
				RoutineCrash(ctx)
			}()
         wg.Wait() //no need for the Eventually
		}, NodeTimeout(time.Second*2))

		It("no crash", func(ctx SpecContext) {
			wg = &sync.WaitGroup{}
			wg.Add(1)
			go func() {
				defer GinkgoRecover()
            defer wg.Done() //always tell the wait group we are done
				RoutineGood(ctx) //take the context so we can cancel early if needed
			}()
         wg.Wait() //no need for the Eventually
		}, NodeTimeout(time.Second*6))
	})
})
```

**Comment by onsi** [comment]
sorry the formatting was messed up - but you should get the basic gist of it.

**Comment by SfnUser** [comment]
Thanks @onsi.

Interesting. Yes, it does work with `NodeTimeout` expiry itself. 

So, the recommended way seems to pass the `ctx` across all the GO routines invoked in the Node and make them listen to `ctx.Done()`. I also see why you have `wg.Done()` as `defer` call because even in the case of runtime violation,  all the deferred calls of the Go routine are executed (that's how GinkgoRecovers) when the context is cancelled on `NodeTimeout` expiry (which is not possible with the previous code) and that's why`wg.Wait()` unblocks. Will note that down.

Thanks for the follow-up. I am closing the issue
